Setting allowsContentJavaScript=NO does not show <noscript> elements

hortont424 opened the pull request:
#### cc0631f6705ab4beef5007a7e7965d4421b7b1b1
<pre>
Setting allowsContentJavaScript=NO does not show &lt;noscript&gt; elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=276037">https://bugs.webkit.org/show_bug.cgi?id=276037</a>
<a href="https://rdar.apple.com/130822453">rdar://130822453</a>

Reviewed by Richard Robinson and Geoffrey Garen.

* Source/WebCore/html/parser/HTMLParserOptions.cpp:
(WebCore::HTMLParserOptions::HTMLParserOptions):
Sites generally expect that if their script won&apos;t run, &lt;noscript&gt; elements
are inserted into the DOM. Some use this, for example, to insert style that
hides elements that won&apos;t be populated by forthcoming changes from script.

Previously, WebKit would insert &lt;noscript&gt; elements if JavaScript was totally
disabled, but not if it was disabled by the &quot;disable content JavaScript&quot; mechanism
(which still allows the user agent to run script, just denies script from the page).

However, given the overall purpose of &lt;noscript&gt; is about script from the page,
it seems prudent to insert &lt;noscript&gt; contents in both cases.

The scriptingFlag in the parser seems to only affect &lt;noscript&gt; participation,
so this minimal patch appears to do the trick without any unexpected downstream
effects.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEvaluateJavaScript.mm:
(TEST(WebKit, AllowsContentJavaScriptAffectsNoscriptElements)):
Add a test that covers this behavior.

Canonical link: <a href="https://commits.webkit.org/280544@main">https://commits.webkit.org/280544@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad68a99a8757405e4284d28c6603fa3987066016

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56918 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36246 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9392 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60538 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7361 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59046 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43869 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7551 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46101 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5171 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58948 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34046 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49126 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26959 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30826 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6457 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6366 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52786 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6728 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62219 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/831 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6832 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53356 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/834 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49181 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53395 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12582 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/707 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32075 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33160 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34245 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32906 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEvaluateJavaScript.mm:
```diff
@@ -485,6 +485,20 @@ HTTPServer server(HTTPServer::UseCoroutines::Yes, [](Connection connection) -> T
 
+TEST(WebKit, AllowsContentJavaScriptAffectsNoscriptElements)
+{
+    RetainPtr<WKWebpagePreferences> preferences = adoptNS([[WKWebpagePreferences alloc] init]);
```
these RetainPtr's don't need the explicit template parameter specified I think

geoffreygaren reviewed (approved):
r=me

That's the beauty of allowsContentJavaScript -- it's a parsing-only concept.

hortont424 committed:
Setting allowsContentJavaScript=NO does not show <noscript> elements
https://bugs.webkit.org/show_bug.cgi?id=276037
rdar://130822453

Reviewed by Richard Robinson and Geoffrey Garen.

* Source/WebCore/html/parser/HTMLParserOptions.cpp:
(WebCore::HTMLParserOptions::HTMLParserOptions):
Sites generally expect that if their script won't run, <noscript> elements
are inserted into the DOM. Some use this, for example, to insert style that
hides elements that won't be populated by forthcoming changes from script.

Previously, WebKit would insert <noscript> elements if JavaScript was totally
disabled, but not if it was disabled by the "disable content JavaScript" mechanism
(which still allows the user agent to run script, just denies script from the page).

However, given the overall purpose of <noscript> is about script from the page,
it seems prudent to insert <noscript> contents in both cases.

The scriptingFlag in the parser seems to only affect <noscript> participation,
so this minimal patch appears to do the trick without any unexpected downstream
effects.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEvaluateJavaScript.mm:
(TEST(WebKit, AllowsContentJavaScriptAffectsNoscriptElements)):
Add a test that covers this behavior.

Canonical link: https://commits.webkit.org/280544@main
